### DISCUSSION TITLE
Handling httpx exceptions in Application.builder based app

### DISCUSSION THREAD

**Discussion by b-ferd** [python-telegram-bot/python-telegram-bot#4175] [Q&A] [answered]
I've toyed around a bit with the error handler example code [errorhandlerbot.py](https://docs.python-telegram-bot.org/en/stable/examples.errorhandlerbot.html), for Application.builder based PTB applications. The code comments seem to indicate that the handler is only called on errors that happen during the registered message handler callback functions, and the [Exception Handling Docu](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Exceptions%2C-Warnings-and-Logging) indicates this as well. On the [Handling Network Errors Page](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Handling-network-errors), there is some hinting towards catching exceptions, but no specific suggestion how to catch exceptions raised by httpx.

With the raw bot, this was easy: a try-catch-block around `bot=Bot()` and around `await bot.get_updates() `.
But how do I catch this with the `Application.builder` based bot? The Network Error info page says:
"_When a network error occurs, be prepared to catch the raised exception and handle it according to your policy ... or use PTBs built-in mechanism for exception handling_." ... well, yes... but how do I catch httpx exceptions when I'm not the caller?

Background: aim is to detect network outage errors asap (i.e. I _don't_ want to tweak connection timeout settings to cover things up), log a brief warning message, recover automatically (PTB already does this)...

Is there already example code for catching httpx exceptions in Application.builder based applications that I missed?
Any help or suggestions, as usual, are very much appreciated.

**Comment by b-ferd** [comment, marked as answer]
Solved itself!
I found this [posting 3430](https://github.com/python-telegram-bot/python-telegram-bot/issues/3430) by Bibo-Joshi.
This is the working code snippet, for reference:
```
def error_callback(error, application):
	application.create_task(application.process_error(update=None, error=error))

async def botloop() -> NoReturn:
	application = Application.builder().token(bot_token).build()
	application.add_handler(CommandHandler("start", start))
	application.add_error_handler(error_handler)
	await application.initialize()
	await application.updater.start_polling(allowed_updates=Update.ALL_TYPES, \
		error_callback=functools.partial(error_callback, application=application))
	await application.start()
	
	while True:
		await asyncio.sleep(1)
	
	# dropped from the loop -> exit. shutdown bot.
	await application.Updater.stop()
	await application.stop()
	await application.shutdown()
```
This will pass network errors from httpx into the error_handler() function, there they can be detected using `if type(context.error) == telegram.error.NetworkError`.

**Reply by Bibo-Joshi** [reply]
Glas you found your answer. Just one hint: `isinstance` is usually considered more pythonic than exactly matching the type and also catches subclasses :)

**Comment by b-ferd** [comment]
btw... suggest to add this info to the Networking Error page...

**Reply by Bibo-Joshi** [reply]
You are welcome to edit the wiki pages where you a seed :)
